pkg/snapshot/restorer/restorer.go

```diff
@@ -887,3 +944,86 @@ func ErrorArrayToError(errs []error) error {
 
 	return fmt.Errorf("%s", strings.TrimSpace(errString))
 }
+
+// HandleAlarm function handles alarm raised by backup-restore.
+func (r *Restorer) HandleAlarm(stopHandleAlarmCh chan bool, dbSizeAlarmCh <-chan string, dbSizeDisAlarmCh chan bool, clientMaintenance client.MaintenanceCloser) {
+	r.logger.Info("Starting to handle an alarm...")
+	for {
+		select {
+		case <-stopHandleAlarmCh:
+			r.logger.Info("Closing handleAlarm...")
+			return
+		case endPoint := <-dbSizeAlarmCh:
+			r.logger.Info("Received a dbsize alarm")
+			r.logger.Infof("Calling defrag on endpoint: [%v]", endPoint)
+			if err := func() error {
+				ctx, cancel := context.WithTimeout(context.Background(), etcdDefragTimeout)
+				defer cancel()
+				if _, err := clientMaintenance.Defragment(ctx, endPoint); err != nil {
+					return err
+				}
+				return nil
+			}(); err != nil {
+				r.logger.Errorf("unable to disalarm as defrag call failed: %v", err)
+				// failed to disalarm
+				dbSizeDisAlarmCh <- false
+			} else {
+				// successfully disalarm
+				dbSizeDisAlarmCh <- true
+			}
+		}
+	}
+}
+
+// MakeEtcdLeanAndCheckAlarm calls etcd compaction on given revision number and raise db size alarm if embedded etcd db size crosses threshold.
+func (r *Restorer) MakeEtcdLeanAndCheckAlarm(revision int64, endPoints []string, embeddedEtcdQuotaBytes float64, dbSizeAlarmCh chan string, dbSizeDisAlarmCh <-chan bool, clientKV client.KVCloser, clientMaintenance client.MaintenanceCloser) error {
+	if err := func() error {
+		ctx, cancel := context.WithTimeout(context.Background(), etcdCompactTimeout)
+		defer cancel()
+		if _, err := clientKV.Compact(ctx, revision, clientv3.WithCompactPhysical()); err != nil {
+			return fmt.Errorf("Compact API call failed: %v", err)
+		}
+		return nil
+	}(); err != nil {
+		return err
```

Comment:
Why is there an anonymous function used here?
It appears that within this function, an error is being handled. Are you expecting the anonymous function to return an error, and then, if an error is present, to return that error from the main function

Reply:
😕

Reply:
> anonymous function used here

anonymous function is used to cancel the ctx using `defer cancel()` , defer is execute at the end of function, hence there was no choice but to go with anonymous function.

> It appears that within this function, an error is being handled. Are you expecting the anonymous function to return an error, and then, if an error is present, to return that error from the main function

please check the whole code, when anonymous function returns error we log that error send the `dbSizeDisAlarmCh <- false` to channel to indicate it fails

Reply:
>please check the whole code, when anonymous function returns error we log that error send the dbSizeDisAlarmCh <- false to channel to indicate it fails

You are referring `HandleAlarm` but here it's not necessary to anonymous function.

if you feel, immediately you want to cancel context in above pointed code, then do it consistently for all the context in this function.

Reply:
> if you feel, immediately you want to cancel context in above pointed code, then do it consistently for all the context in this function.

ideally it should be but it depends on context timeout like status call have context.Timeout every small as compared to compaction and defragmentation.

Reply:
If you want then I can make these changes.

Reply:
ok, If you aren't satisfied, then I will make the changes as you requested. I hope that's fine.
But that shouldn't be bigger issue as we discussed both ways are doing similar thing.

Reply:
I have removed the anonymous function from compaction. PTAL